internal/controllers/operator_controller_test.go

```diff
@@ -32,1028 +30,1234 @@ import (
-							Expect(cond).NotTo(BeNil())
-							Expect(cond.Status).To(Equal(metav1.ConditionFalse))
-							Expect(cond.Reason).To(Equal(operatorsv1alpha1.ReasonInstallationFailed))
-							Expect(cond.Message).To(Equal("bundledeployment not ready: installing"))
-						})
-
-						It("verify operator status when bundleDeployment installation status is unknown", func() {
-							apimeta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-								Type:    rukpakv1alpha1.TypeInstalled,
-								Status:  metav1.ConditionUnknown,
-								Message: "installing",
-								Reason:  rukpakv1alpha1.ReasonInstallationSucceeded,
-							})
-
-							By("updating the status of bundleDeployment")
-							err := cl.Status().Update(ctx, bd)
-							Expect(err).NotTo(HaveOccurred())
-
-							By("running reconcile")
-							res, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: opKey})
-							Expect(res).To(Equal(ctrl.Result{}))
-							Expect(err).NotTo(HaveOccurred())
-
-							By("fetching the updated operator after reconcile")
-							op := &operatorsv1alpha1.Operator{}
-							err = cl.Get(ctx, opKey, op)
-							Expect(err).NotTo(HaveOccurred())
-
-							By("Checking the status fields")
-							Expect(op.Status.ResolvedBundleResource).To(Equal("quay.io/operatorhubio/prometheus@fake2.0.0"))
-							Expect(op.Status.InstalledBundleResource).To(Equal(""))
-
-							By("checking the expected conditions")
-							cond := apimeta.FindStatusCondition(op.Status.Conditions, operatorsv1alpha1.TypeResolved)
-							Expect(cond).NotTo(BeNil())
-							Expect(cond.Status).To(Equal(metav1.ConditionTrue))
-							Expect(cond.Reason).To(Equal(operatorsv1alpha1.ReasonSuccess))
-							Expect(cond.Message).To(Equal("resolved to \"quay.io/operatorhubio/prometheus@fake2.0.0\""))
-							cond = apimeta.FindStatusCondition(op.Status.Conditions, operatorsv1alpha1.TypeInstalled)
-							Expect(cond).NotTo(BeNil())
-							Expect(cond.Status).To(Equal(metav1.ConditionFalse))
-							Expect(cond.Reason).To(Equal(operatorsv1alpha1.ReasonInstallationFailed))
-							Expect(cond.Message).To(Equal("bundledeployment not ready: installing"))
-						})
-
-					})
-
-				})
-			})
-			When("an out-of-date BundleDeployment exists", func() {
-				var bd *rukpakv1alpha1.BundleDeployment
-				BeforeEach(func() {
-					By("creating the expected BD")
-					bd = &rukpakv1alpha1.BundleDeployment{
-						ObjectMeta: metav1.ObjectMeta{Name: opKey.Name},
-						Spec: rukpakv1alpha1.BundleDeploymentSpec{
-							ProvisionerClassName: "foo",
-							Template: &rukpakv1alpha1.BundleTemplate{
-								Spec: rukpakv1alpha1.BundleSpec{
-									ProvisionerClassName: "bar",
-									Source: rukpakv1alpha1.BundleSource{
-										Type: rukpakv1alpha1.SourceTypeHTTP,
-										HTTP: &rukpakv1alpha1.HTTPSource{
-											URL: "http://localhost:8080/",
-										},
-									},
-								},
-							},
+func TestOperatorBundleDeploymentOutOfDate(t *testing.T) {
```

Comment:
Having problems here, as the tests appear to be stomping on each other since they are sharing some resources. This will require more than a simple adding of `t.Parallel()`.

Reply:
Sounds like good effort to invest :)

Reply:
As I'm working through this, it's increasing test time from ~7s to ~35 seconds, likely due to the additional test environments being created...

Reply:
You can have a single shared environment (wasn't that what the TestMain was doing) while running tests in parallel

Reply:
But they stomp on each other with the shared environment.
I have completely parallelized it, but that increases the test time.
Trying to figure out the stomping will likely result in a better test, but it may take _a lot_ more effort to figure it out.

Admittedly, 7 seconds is not a lot of time for the serialized test...

Reply:
Mind pasting what you mean by "stomping"? You randomized all of the object names, I am surprised there is still cross-chatter.

Reply:
This is the latest. Randomized names are used. The tests have to be evaluated for their ability to run with others. As with OLMv0 tests, these tests use common packages, etc, for testing, so one test could interfere with another. If not run in parallel, or with separate test environments, these pass.
```
--- FAIL: TestOperatorBundleDeploymentUpToDate (0.04s)
    operator_controller_test.go:286: When the operator specifies a valid available package
    operator_controller_test.go:287: By initializing cluster state
    operator_controller_test.go:294: When the expected BundleDeployment already exists
    operator_controller_test.go:295: When the BundleDeployment spec is up-to-date
    operator_controller_test.go:296: By patching the existing BD
    operator_controller_test.go:330: When the BundleDeployment status is mapped to the expected Operator status
    operator_controller_test.go:331: It verifies operator status when bundle deployment is waiting to be created
    operator_controller_test.go:332: By updating the status of bundleDeployment
    operator_controller_test.go:333:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:333
                Error:          Received unexpected error:
                                Operation cannot be fulfilled on bundledeployments.core.rukpak.io "operator-test-mbpf6rvf": StorageError: invalid object, Code: 4, Key: /registry/core.rukpak.io/bundledeployments/operator-test-mbpf6rvf, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 5fa2cdb4-31e8-48a3-b171-3422a71ab04c, UID in object meta:
                Test:           TestOperatorBundleDeploymentUpToDate
--- FAIL: TestOperatorChannelVersionExists (0.04s)
    operator_controller_test.go:689: When the operator specifies a channel with version that exist
    operator_controller_test.go:690: By initializing cluster state
    operator_controller_test.go:705: It sets resolution success status
    operator_controller_test.go:706: By running reconcile
    operator_controller_test.go:709:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:709
                Error:          Received unexpected error:
                                [no package "non-existent-szblph" found, Operation cannot be fulfilled on operators.operators.operatorframework.io "operator-test-cdflnjps": StorageError: invalid object, Code: 4, Key: /registry/operators.operatorframework.io/operators/operator-test-cdflnjps, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 519c027b-3f77-40ae-a0b7-b8c46b88e573, UID in object meta: ]
                Test:           TestOperatorChannelVersionExists
--- FAIL: TestOperatorBundleDeploymentDoesNotExist (0.04s)
    operator_controller_test.go:137: When the operator specifies a valid available package
    operator_controller_test.go:138: By initializing cluster state
    operator_controller_test.go:145: When the BundleDeployment does not exist
    operator_controller_test.go:146: By running reconcile
    operator_controller_test.go:149:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:149
                Error:          Received unexpected error:
                                no package "non-existent-szblph" found
                Test:           TestOperatorBundleDeploymentDoesNotExist
--- FAIL: TestOperatorBadBundleMediaType (0.04s)
    operator_controller_test.go:1020: When the operator specifies a package with a bad bundle mediatype
    operator_controller_test.go:1021: By initializing cluster state
    operator_controller_test.go:1035: It sets resolution success status
    operator_controller_test.go:1036: By running reconcile
    operator_controller_test.go:1040:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:1040
                Error:          Error "no package \"non-existent-szblph\" found" does not contain "unknown bundle mediatype: badmedia+v1"
                Test:           TestOperatorBadBundleMediaType
--- FAIL: TestOperatorChannelExistsNoVersion (0.05s)
    operator_controller_test.go:750: When the operator specifies a package that exists within a channel but no version specified
    operator_controller_test.go:751: By initializing cluster state
    operator_controller_test.go:765: It sets resolution success status
    operator_controller_test.go:766: By running reconcile
    operator_controller_test.go:769:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:769
                Error:          Received unexpected error:
                                no package "non-existent-szblph" found
                Test:           TestOperatorChannelExistsNoVersion
--- FAIL: TestOperatorNoVersion (0.05s)
    operator_controller_test.go:910: When the operator specifies a package version that does not exist in the channel
    operator_controller_test.go:911: By initializing cluster state
    operator_controller_test.go:925: It sets resolution failure status
    operator_controller_test.go:926: By running reconcile
    operator_controller_test.go:929:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:929
                Error:          Error message not equal:
                                expected: "no package \"prometheus\" matching version \"0.57.0\" found in channel \"non-existent\""
                                actual  : "no package \"non-existent-szblph\" found"
                Test:           TestOperatorNoVersion
--- FAIL: TestOperatorVersionNoChannel (0.05s)
    operator_controller_test.go:809: When the operator specifies a package version in a channel that does not exist
    operator_controller_test.go:810: By initializing cluster state
    operator_controller_test.go:824: It sets resolution failure status
    operator_controller_test.go:825: By running reconcile
    operator_controller_test.go:828:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:828
                Error:          Error message not equal:
                                expected: "no package \"prometheus\" matching version \"0.47.0\" found in channel \"alpha\""
                                actual  : "no package \"non-existent-szblph\" found"
                Test:           TestOperatorVersionNoChannel
--- FAIL: TestOperatorNoChannel (0.05s)
    operator_controller_test.go:861: When the operator specifies a package in a channel that does not exist
    operator_controller_test.go:862: By initializing cluster state
    operator_controller_test.go:874: It sets resolution failure status
    operator_controller_test.go:875: By running reconcile
    operator_controller_test.go:878:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:878
                Error:          Error message not equal:
                                expected: "no package \"prometheus\" found in channel \"non-existent\""
                                actual  : "no package \"non-existent-szblph\" found"
                Test:           TestOperatorNoChannel
--- FAIL: TestOperatorDuplicatePackage (0.05s)
    operator_controller_test.go:638: When the operator specifies a duplicate package
    operator_controller_test.go:639: By initializing cluster state
    operator_controller_test.go:652: It sets resolution failure status
    operator_controller_test.go:653: By running reconcile
    operator_controller_test.go:656:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:656
                Error:          Error message not equal:
                                expected: "duplicate identifier \"required package prometheus\" in input"
                                actual  : "no package \"non-existent-szblph\" found"
                Test:           TestOperatorDuplicatePackage
--- FAIL: TestOperatorNonExistantVersion (0.05s)
    operator_controller_test.go:88: When the operator specifies a version that does not exist
    operator_controller_test.go:89: By initializing cluster state
    operator_controller_test.go:100: It sets resolution failure status
    operator_controller_test.go:101: By running reconcile
    operator_controller_test.go:104:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:104
                Error:          Error message not equal:
                                expected: "no package \"prometheus\" matching version \"0.50.0\" found"
                                actual  : "no package \"non-existent-szblph\" found"
                Test:           TestOperatorNonExistantVersion
--- FAIL: TestOperatorPlainV0Bundle (0.05s)
    operator_controller_test.go:961: When the operator specifies a package with a plain+v0 bundle
    operator_controller_test.go:962: By initializing cluster state
    operator_controller_test.go:976: It sets resolution success status
    operator_controller_test.go:977: By running reconcile
    operator_controller_test.go:980:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:980
                Error:          Received unexpected error:
                                no package "non-existent-szblph" found
                Test:           TestOperatorPlainV0Bundle
--- FAIL: TestOperatorExpectedBundleDeployment (0.05s)
    operator_controller_test.go:562: When the operator specifies a valid available package
    operator_controller_test.go:563: By initializing cluster state
    operator_controller_test.go:570: When an out-of-date BundleDeployment exists
    operator_controller_test.go:571: By creating the expected BD
    operator_controller_test.go:591: By running reconcile
    operator_controller_test.go:594:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:594
                Error:          Received unexpected error:
                                no package "non-existent-szblph" found
                Test:           TestOperatorExpectedBundleDeployment
--- FAIL: TestOperatorBundleDeploymentOutOfDate (0.05s)
    operator_controller_test.go:194: When the operator specifies a valid available package
    operator_controller_test.go:195: By initializing cluster state
    operator_controller_test.go:202: When the expected BundleDeployment already exists
    operator_controller_test.go:203: When the BundleDeployment spec is out of date
    operator_controller_test.go:204: By patching the existing BD
    operator_controller_test.go:235: By modifying the BD spec and creating the object
    operator_controller_test.go:239: It results in the expected BundleDeployment
    operator_controller_test.go:241: By running reconcile
    operator_controller_test.go:244:
                Error Trace:    /home/tshort/git/operator-framework/operator-controller/internal/controllers/operator_controller_test.go:244
                Error:          Received unexpected error:
                                no package "non-existent-szblph" found
                Test:           TestOperatorBundleDeploymentOutOfDate
FAIL

Reply:
The packageNames are more difficult to randomize; it might require each test to have it's own set of test data, which is currently shared.

Reply:
jeeez .... let's worry about parallel later then., ugh

Reply:
I'm in the middle of an update that might just work... So, let's hold off at least until Thursday.

Reply:
It looks as though the reconcilers are processing interleaved requests or something? Getting errors in test A that are for test B and errors in test C that are for test A, etc. Weird.

Reply:
If it's from resolution, I think that's because resolution is totally global right now - if it's for other things, that seems like some bug you found

Reply:
Yes, that appears to be the case.